llvm/lib/Target/RISCV/Disassembler/RISCVDisassembler.cpp

```diff
@@ -341,6 +341,19 @@ static DecodeStatus decodeUImmOperandGE(MCInst &Inst, uint32_t Imm,
   return MCDisassembler::Success;
 }
 
+template <unsigned Width, unsigned LowerBound>
+static DecodeStatus decodeUImmPlus1OperandGE(MCInst &Inst, uint32_t Imm,
+                                             int64_t Address,
+                                             const MCDisassembler *Decoder) {
+  assert(isUInt<Width>(Imm) && "Invalid immediate");
+
+  if (Imm < LowerBound)
+    return MCDisassembler::Fail;
```

Comment:
I find this confusing, and I think how this is used is incorrect.

The `uin32_t Imm` argument to this function will be the 5-bit field out of the instruction encoding (with the bits in the right place if they were moved around). This method is for taking the bits of that encoding, and turning it back into the correct value (or failing to disassemble).

In this case, I think you want the code to be the following, instead, because this gives the Lower Bound in terms of the decoded immediate (which we prefer to match the value the instruction aims to use), rather than in terms of the encoded value:

```suggestion
  if ((Imm + 1) < LowerBound)
    return MCDisassembler::Fail;
```

I think we should have a test to check this case. Add one to `llvm/test/MC/Disassembler/RISCV/xqci-invalid.txt`, where you hand encode a list of bytes. The case you want to hit is where the encoding would be rejected - I think in this case where the encoding value of `width_minus1` is equal to 4.

Reply:
Done. Please have a look at the test and let me know if I should change anything.